serverless-architecture-on-aws/lab-3/README.md

```diff
@@ -0,0 +1,459 @@
+    managedPolicyArns: [
+        aws.iam.ManagedPolicy.AWSLambdaBasicExecutionRole, 
+    ]
+});
+
+const fn = new aws.lambda.Function("function", {
+    role: role.arn,
+    code: new pulumi.asset.AssetArchive({
+        ".": new pulumi.asset.FileArchive("./app")
+    }),
+    handler: "s3.handler",
+    runtime: aws.lambda.Runtime.NodeJS16dX
+});
+
+const logGroup = new aws.cloudwatch.LogGroup("loggroup", {
+    name: pulumi.interpolate`/aws/lambda/${fn.name}`
+});
+
+const bucket = new aws.s3.Bucket("lambdaBucket");
+bucket.onObjectCreated("onObjectCreated", fn);
+
+export const bucketName = bucket.bucket;
+```
+</details>
+
+### CloudWatch Schedule
+
+If you've not used this before it's very useful because you can effectively create CRON jobs that run your Lambda Functions.
+
+Same as before, here's the Lambda Function code:
+
+```javascript
+exports.handler =  async function(event, context) {
+    const cloudwatchEvent = event;
+    const logJSON = {
+        id: event.id,
+        timestamp: event.time,
+        eventArn: event.resources[0]
+    };
+    console.log(logJSON)
+}
+```
+
+Here's the Lambda Function resource with the handler input changed:
+
+```typescript
+const fn = new aws.lambda.Function("function", {
+    role: role.arn,
+    code: new pulumi.asset.AssetArchive({
+        ".": new pulumi.asset.FileArchive("./app")
+    }),
+    handler: "cloudwatch.handler", // changed from "s3.handler"
+    runtime: aws.lambda.Runtime.NodeJS16dX
+});
+```
+
+The code for the CloudWatch Scheduler resource:
+
+```typescript
+const scheduler = new aws.cloudwatch.EventRule("scheduler", {
+    scheduleExpression: "rate(1 minute)"
+});
+```
+
+and to wire it up to the Lambda Function:
+
+```typescript
+scheduler.onEvent("schedulerevent", fn)
+```
+
+As we're not going to use the CLI to run the scheduler, we don't need to export any outputs.
+
+Then an update:
+
+```zsh
+pulumi up
+```
+
+<details><summary>The complete CloudWatch Scheduler Pulumi code can be seen here</summary>
+
+```typescript
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+
+const role = new aws.iam.Role("role", {
+    assumeRolePolicy: aws.iam.assumeRolePolicyForPrincipal(aws.iam.Principals.LambdaPrincipal),
+    managedPolicyArns: [
+        aws.iam.ManagedPolicy.AWSLambdaBasicExecutionRole, 
+    ]
+});
+
+const fn = new aws.lambda.Function("function", {
+    role: role.arn,
+    code: new pulumi.asset.AssetArchive({
+        ".": new pulumi.asset.FileArchive("./app")
+    }),
+    handler: "cloudwatch.handler",
+    runtime: aws.lambda.Runtime.NodeJS16dX
+});
+
+const logGroup = new aws.cloudwatch.LogGroup("loggroup", {
+    name: pulumi.interpolate`/aws/lambda/${fn.name}`
+});
+
+const scheduler = new aws.cloudwatch.EventRule("scheduler", {
+    scheduleExpression: "rate(1 minute)"
+});
+
+scheduler.onEvent("schedulerevent", fn)
+```
+</details>
+
+## Tidying up
```

Comment:
Rename title:  Destroy the Resources - Critical Step to avoid cloud charges

Reply:
Everything that I deploy is covered under the free tier (as I say below), and as it's serverless you don't get charged unless you actually run things so I'm going to leave this.